Add six Krylov processes (#618)
JuliaSmoothOptimizers/Krylov.jl: docs/src/processes.md

# [Krylov processes](@id krylov-processes)

> Krylov processes are the foundations of Krylov methods, they generate bases of Krylov subspaces.

Reviewer: ```suggestion
Krylov processes are the foundation of Krylov methods, they generate bases of Krylov subspaces.
```

### Notation

Define $V_k := \begin{bmatrix} v_1 & \ldots & v_k \end{bmatrix} \enspace$ and $\enspace U_k := \begin{bmatrix} u_1 & \ldots & u_k \end{bmatrix}$.

For a matrix $C \in \mathbb{C}^{n \times n}$ and a vector $t \in \mathbb{C}^{n}$, the $k$-th Krylov subspace generated by $C$ and $t$ is
```math
\mathcal{K}_k(C, t) :=
\left\{\sum_{i=0}^{k-1} \omega_i C^i t \, \middle \vert \, \omega_i \in \mathbb{C},~0 \le i \le k-1 \right\}.
```

For matrices $C \in \mathbb{C}^{n \times n} \enspace$ and $\enspace T \in \mathbb{C}^{n \times p}$, the $k$-th block Krylov subspace generated by $C$ and $T$ is
```math
\mathcal{K}_k^{\square}(C, T) :=
\left\{\sum_{i=0}^{k-1} C^i T \, \Omega_i \, \middle \vert \, \Omega_i \in \mathbb{C}^{p \times p},~0 \le i \le k-1 \right\}.
```

> ## Symmetric Lanczos

Author: ```suggestion
## Hermitian Lanczos
```

> ![symmetric_lanczos](./graphics/symmetric_lanczos.png)

Reviewer: Out of curiosity, does MathJax allow us to typeset algorithms?

Reviewer: Also, specify that $A = A^H$.

Author: > Out of curiosity, does MathJax allow us to typeset algorithms?

I will [pseudocode.js](https://github.com/SaswatPadhi/pseudocode.js).

> After $k$ iterations of the symmetric Lanczos process, the situation may be summarized as

Reviewer: we should really call it the Hermitian Lanczos process.

> After $k$ iterations of the symmetric Lanczos process, the situation may be summarized as

Author: ```suggestion
After $k$ iterations of the Hermitian Lanczos process, the situation may be summarized as
```

```math
\begin{align*}
  A V_k &= V_k T_k + \beta_{k+1,k} v_{k+1} e_k^T = V_{k+1}  T_{k+1,k}, \\
  V_k^H V_k &= I_k,
\end{align*}
```
where $V_k$ is an orthonormal basis of the Krylov subspace $\mathcal{K}_k (A,b)$,
```math
T_k =
\begin{bmatrix}
  \alpha_1 & \beta_2  &          &         \\
  \beta_2  & \alpha_2 & \ddots   &         \\
           & \ddots   & \ddots   & \beta_k \\
           &          & \beta_k  & \alpha_k

> \end{bmatrix}

Reviewer: I think T should be Hermitian here.

Author: T has always real coefficients, it is symmetric.

, \qquad
T_{k+1,k} =
\begin{bmatrix}
  T_{k} \\
  \beta_{k+1} e_{k}^T
\end{bmatrix}.
```

The function [`symmetric_lanczos`](@ref symmetric_lanczos) returns $V_{k+1}$ and $T_{k+1,k}$.

Related methods: [`SYMMLQ`](@ref symmlq), [`CG`](@ref cg), [`CR`](@ref cr), [`MINRES`](@ref minres), [`MINRES-QLP`](@ref minres_qlp), [`CGLS`](@ref cgls), [`CRLS`](@ref crls), [`CGNE`](@ref cgne), [`CRMR`](@ref crmr), [`CG-LANCZOS`](@ref cg_lanczos) and [`CG-LANCZOS-SHIFT`](@ref cg_lanczos_shift).

```@docs
symmetric_lanczos
```

> ## Unsymmetric Lanczos

Reviewer: This is another bad name in the complex case. Maybe Non-Hermitian Lanczos?

> ## Unsymmetric Lanczos

Author: ```suggestion
## Non-Hermitian Lanczos
```

![unsymmetric_lanczos](./graphics/unsymmetric_lanczos.png)

> After $k$ iterations of the unsymmetric Lanczos process (also named the Lanczos biorthogonalization process), the situation may be summarized as

Author: ```suggestion
After $k$ iterations of the non-Hermitian Lanczos process (also named the Lanczos biorthogonalization process), the situation may be summarized as
```

> ### Notations

Reviewer 2: Similar to the other pages in the doc, I think it misses `Krylov processes` as a title followed by one sentence explaining what is here.

![saunders_simon_yip](./graphics/saunders_simon_yip.png)

After $k$ iterations of the Saunders-Simon-Yip process (also named the orthogonal tridiagonalization process), the situation may be summarized as
```math
\begin{align*}
  A U_k &= V_k T_k + \beta_{k+1}  v_{k+1} e_k^T = V_{k+1} T_{k+1,k}, \\
  B V_k &= U_k S_k + \gamma_{k+1} u_{k+1} e_k^T = U_{k+1} S_{k+1,k}, \\
  V_k^T V_k &= U_k^T U_k = I_k,
\end{align*}
```
where $\begin{bmatrix} V_k & 0 \\ 0 & U_k \end{bmatrix}$ is an orthonormal basis of the block Krylov subspace $\mathcal{K}^{\square}_k \left(\begin{bmatrix} 0 & A \\ A^T & 0 \end{bmatrix}, \begin{bmatrix} b & 0 \\ 0 & c \end{bmatrix}\right)$,
```math
T_k = 
\begin{bmatrix}
  \alpha_1 & \gamma_2 &          &          \\
  \beta_2  & \alpha_2 & \ddots   &          \\
           & \ddots   & \ddots   & \gamma_k \\
           &          & \beta_k  & \alpha_k
\end{bmatrix}
, \qquad
S_k =
\begin{bmatrix}
  \bar{\alpha}_1 & \beta_2        &          &               \\
  \gamma_2       & \bar{\alpha}_2 & \ddots   &               \\
                 & \ddots         & \ddots   & \beta_k       \\
                 &                & \gamma_k  & \bar{\alpha}_k

> \end{bmatrix}

Author: ```suggestion
\end{bmatrix},
```
